### PR TITLE
created obs FM-104

### PR DESCRIPTION
**Pull request by sashrika**
Initial commit. 

https://issues.openmrs.org/browse/FM-104.

From this I could create a Obs with the major attributes. 
I came accross with a issue. The mesuring units that are in the observation and the units that we have put in the "valueQuantity" of the representation can be different. So do we need to handle the unit conversions? 

@harsha89 I formatted the code. Please check is it ok. 

Still I haven't written the test method for this.

### CODE REVIEW

File: api/src/main/java/org/openmrs/module/fhir/api/util/FHIRObsUtil.java [hunk 1]
```diff
@@ -240,4 +243,47 @@ public static Observation generateObs(Obs obs) {
...
+		
+		DateTimeDt dateApplies = (DateTimeDt) observation.getApplies();
+		obs.setObsDatetime(dateApplies.getValue());
+		
+		Date instant = observation.getIssued();
+		obs.setDateCreated(instant);
+		
+		CodeableConceptDt dt = observation.getCode();
+		List<CodingDt> dts = dt.getCoding();
+		CodingDt coding = dts.get(0);
```
Comment (sashrika) [inline api/src/main/java/org/openmrs/module/fhir/api/util/FHIRObsUtil.java]: if the dts is null, we should abort the operation

File: api/src/main/java/org/openmrs/module/fhir/api/util/FHIRObsUtil.java [hunk 2]
```diff
@@ -240,4 +243,47 @@ public static Observation generateObs(Obs obs) {
...
+		String conceptUuid = coding.getCode();
+		Concept concept = Context.getConceptService().getConceptByUuid(conceptUuid);
+		obs.setConcept(concept);
+		
+		if (concept.isNumeric()) {
+			QuantityDt quantity = (QuantityDt) observation.getValue();
+			BigDecimal bd = quantity.getValue();
+			double doubleValue = bd.doubleValue();
+			obs.setValueNumeric(doubleValue);
+		}
```
Comment (sashrika) [inline api/src/main/java/org/openmrs/module/fhir/api/util/FHIRObsUtil.java]: Here, will we need a unit conversion?

File: api/src/main/java/org/openmrs/module/fhir/api/ObsService.java [hunk 3]
```diff
@@ -91,9 +91,16 @@
 	/**
 	 * Delete observation by id
 	 *
-	 * @param id uuid of the observation
-	 * @return observation fhir resource
+	 * @param id uuid of the observation 
+	 * @return observation fhir resource    // RETRUN??? THIS IS A VOID RESOURCE
```
Comment (surangak) [inline api/src/main/java/org/openmrs/module/fhir/api/ObsService.java]: Good catch, please edit this comment :)

File: omod/src/main/java/org/openmrs/module/fhir/providers/RestfulObservationResourceProvider.java [hunk 4]
```diff
@@ -146,5 +146,21 @@ public Observation getResourceById(@IdParam IdDt theId) {
...
+	 *
+	 * @param observation fhir observation object
+	 */
+	@Create
+	public MethodOutcome createFHIRObservation(@ResourceParam Observation observation) {
+		observation = provider.createFHIRObservation(observation);
+		MethodOutcome retVal = new MethodOutcome();
+		retVal.setId(new IdDt(FHIRConstants.OBSERVATION, observation.getId().getIdPart()));
+		OperationOutcome outcome = new OperationOutcome();
+		outcome.addIssue().setDetails("Observation is successfully created"+observation.getId().getIdPart());
```
Comment (surangak) [inline omod/src/main/java/org/openmrs/module/fhir/providers/RestfulObservationResourceProvider.java]: A space might help :)
...created<SPACE>"+observation.getId()

**Comment by surangak** [comment]
@sashrika in regard to your comment on measuring units. If I understood right, your question is, what to do if we try to persist an Obs with LOINC code 12345 which uses the measure unit CM, but the same concept that we have in the OpenMRS concept dictionary uses the measure unit of M, right? Hmm... I don't think that this issue will occur in a practical scenario... and if it did, it would mean that the client or OpenMRS would be wrong... I think it makes sense to ignore such a situation... unless @harsha89  has any suggestions... :)

File: api/src/main/java/org/openmrs/module/fhir/api/impl/ObsServiceImpl.java [hunk 5]
```diff
@@ -223,5 +226,23 @@ public void deleteObs(String id) {
...
 		Context.getObsService().voidObs(obs, "DELETED by FHIR Request");
 	}
+	
+	/**
+	 * @see org.openmrs.module.fhir.api.ObsService#createFHIRObservation(Observation)
+	 */
+	@Override
+	public Observation createFHIRObservation(Observation observation){		
+		List<String> errors = new ArrayList<String>();
+		Obs obs=FHIRObsUtil.generateOpenMRSObs(observation,errors);
```
Comment (harsha89) [inline api/src/main/java/org/openmrs/module/fhir/api/impl/ObsServiceImpl.java]: Again formatting spaces between = and after , :-1:

File: api/src/main/java/org/openmrs/module/fhir/api/impl/ObsServiceImpl.java [hunk 6]
```diff
@@ -223,5 +226,23 @@ public void deleteObs(String id) {
...
+		if (!errors.isEmpty()) {
+			StringBuilder errorMessage = new StringBuilder("The request cannot be processed due to the following issues \n");
+			for (int i = 0; i < errors.size(); i++) {
+				errorMessage.append((i + 1) + " : " + errors.get(i) + "\n");
+			}
+			throw new UnprocessableEntityException(errorMessage.toString());
+		}
+		obs=Context.getObsService().saveObs(obs,"CREATED by FHIR Request");
```
Comment (harsha89) [inline api/src/main/java/org/openmrs/module/fhir/api/impl/ObsServiceImpl.java]: same here

File: api/src/main/java/org/openmrs/module/fhir/api/util/FHIRObsUtil.java (hunk 1 above)
Comment (surangak) [inline api/src/main/java/org/openmrs/module/fhir/api/util/FHIRObsUtil.java]: Agreed. But as per the spec, the Observation resource will always have a code.If it doesn't, then we need to fail, and return an error.

File: api/src/main/java/org/openmrs/module/fhir/api/util/FHIRObsUtil.java (hunk 2 above)
Comment (surangak) [inline api/src/main/java/org/openmrs/module/fhir/api/util/FHIRObsUtil.java]: Um.... I think I addressed this in that email that you sent me?

**Comment by surangak** [comment]
At this point, the pull request deals with numeric values only. I think its a fine first pass, and that we should merge it once @sashrika addresses the comments listed here. Having it in master will help @milankarunarathne proceed further too.

File: api/src/main/java/org/openmrs/module/fhir/api/util/FHIRObsUtil.java [hunk 7]
```diff
@@ -240,4 +241,51 @@ public static Observation generateObs(Obs obs) {
...
+	
+	public static Obs generateOpenMRSObs(Observation observation, List<String> errors) {
+		Obs obs = new Obs();
+		
+		obs.setComment(observation.getComments());
+		if (observation.getSubject() != null) {
+			ResourceReferenceDt subjectref = observation.getSubject();
+			IdDt id = subjectref.getReference();
+			String uri = id.getValue();
+			String[] paths = uri.split("/");
```
Comment (harsha89) [inline api/src/main/java/org/openmrs/module/fhir/api/util/FHIRObsUtil.java]: do we need to split. I mean isn't there any method to get the uuid of method directly using a available method? If not this is fine.

File: api/src/main/java/org/openmrs/module/fhir/api/util/FHIRObsUtil.java [hunk 8]
```diff
@@ -240,4 +241,51 @@ public static Observation generateObs(Obs obs) {
...
+			String patientUuid = paths[1];
+			obs.setPerson(Context.getPersonService().getPersonByUuid(patientUuid));
+		} else {
+			errors.add("Subject cannot be null");
```
Comment (harsha89) [inline api/src/main/java/org/openmrs/module/fhir/api/util/FHIRObsUtil.java]: We are adding error message.Should we return from here?

**Comment by harsha89** [comment]
First version seems to be Ok.Please reply to my comments. I'll merge the pull request.

**Comment by sashrika** [comment]
@surangak  @harsha89  Thanks for the comments. I'll address this today. Yesterday I spend writing test method for this. But surprisingly it fails to save a Obs. I tried with many ways. I don't know why only test method fails :) I have pushed that also. If you see any mistake on that code, please point out :)

And can you have a look at the issue that i have posted in https://mail.google.com/mail/u/1/#inbox/14dc00404bd68df9.

Thank you!

File: api/src/main/java/org/openmrs/module/fhir/api/util/FHIRObsUtil.java (hunk 7 above)
Comment (sashrika) [inline api/src/main/java/org/openmrs/module/fhir/api/util/FHIRObsUtil.java]: What we get is some thing like this 
"encounter": {
"reference": "Encounter/bf51ffc1-1691-11df-97a5-7038c432aabf"
},

As per my knowledge there is no any existing method.

File: api/src/main/java/org/openmrs/module/fhir/api/impl/ObsServiceImpl.java (hunk 5 above)
Comment (sashrika) [inline api/src/main/java/org/openmrs/module/fhir/api/impl/ObsServiceImpl.java]: @surangak  @harsha89  Oops I have edited that after running the formatter. My formatter doesn't run when I build the code. So I set it to run when I save the code every time. Java --> Editor --> Save Actions

Here after this mistakes will not happen ;)

File: api/src/main/java/org/openmrs/module/fhir/api/util/FHIRObsUtil.java (hunk 8 above)
Comment (sashrika) [inline api/src/main/java/org/openmrs/module/fhir/api/util/FHIRObsUtil.java]: I also thought about that. I get this from location resource. In a way, this is good. we accumulate all the erros, then at the serviceImpl method, we check whether the error list is empty or not. If it is not empty we populate ALL error methods to the end user and return unprocessableentitiy exception. 

This way the user get all the errror messages in his request. What do you think?

File: api/src/main/java/org/openmrs/module/fhir/api/util/FHIRObsUtil.java (hunk 1 above)
Comment (sashrika) [inline api/src/main/java/org/openmrs/module/fhir/api/util/FHIRObsUtil.java]: @surangak can you put a link on  where the specification says that the code is always will be there?

File: api/src/main/java/org/openmrs/module/fhir/api/util/FHIRObsUtil.java (hunk 2 above)
Comment (sashrika) [inline api/src/main/java/org/openmrs/module/fhir/api/util/FHIRObsUtil.java]: Yes, it is there.

File: api/src/main/java/org/openmrs/module/fhir/api/util/FHIRObsUtil.java (hunk 8 above)
Comment (harsha89) [inline api/src/main/java/org/openmrs/module/fhir/api/util/FHIRObsUtil.java]: Sounds Okey :)

File: api/src/main/java/org/openmrs/module/fhir/api/util/FHIRObsUtil.java (hunk 7 above)
Comment (harsha89) [inline api/src/main/java/org/openmrs/module/fhir/api/util/FHIRObsUtil.java]: You should be able to use subjectref.getIdPart(). Please fix this and will merge the pull

**Comment by sashrika** [comment]
@harsha89  @surangak  please refer https://github.com/openmrs/openmrs-module-fhir/pull/63 and close this pull req.

**Pull request closed without merging**